Crossbar

ashen0xa opened the pull request:
Initial implementation of TileLink bus arbiter and crossbar.

CircuitCoder committed:
WIP: crossbar

sequencer commented on tilelink/src/xbar/TLArbiter.scala:
```diff
@@ -0,0 +1,110 @@
+            assert(valid === valids)
+            val mask    = RegInit(((BigInt(1) << width) - 1).U(width - 1, 0))
+            val filter  = Cat(valid & (~mask).asUInt, valid)
+            val unready = (rightOR(filter, width * 2, width) >> 1).asUInt | (mask << width).asUInt
```
See https://github.com/chipsalliance/chisel3/pull/2407

sequencer commented:
`LICENSE.SiFive` can be removed, since it's also the Apache-2.0 license, which is duplicated to  the `SPDX-License-Identifier`. please add copyright under the `SPDX-License-Identifier` license.

sequencer commented:
See https://www.sifive.com/documentation, please update to latest TileLink Spec.(CMO in Hint)

ashen0xa committed:
[rtl] implementation of bus arbiter

ashen0xa committed:
[rtl] initial implementation of crossbar

sequencer commented on tilelink/src/bundle/TLLinkParameter.scala:
```diff
@@ -0,0 +1,76 @@
+  def channelAParameter: TileLinkChannelAParameter         =
+    TileLinkChannelAParameter(addressWidth, sourceWidth, dataWidth, sizeWidth)
+  def channelBParameter: Option[TileLinkChannelBParameter] =
+    if (hasBCEChannels)
```
`Option.when`

sequencer commented on tilelink/src/bundle/MessageTypes.scala:
```diff
@@ -0,0 +1,109 @@
+
+/** opcode field of TileLink messages
+  *
+  * cf. TileLink Spec 1.8.1 Table 13
```
plz Refer to correct spec and table

sequencer commented on tilelink/src/xbar/TLArbiter.scala:
```diff
@@ -0,0 +1,110 @@
+            val mask = RegInit(((BigInt(1) << width) - 1).U(width - 1, 0))
+            val filter = Cat(scanRightOr(valid & ~mask), valid)
+            val unready = (filter >> 1).asUInt | (mask << width).asUInt
+            val readys = (~((unready >> width).asUInt & unready(width - 1, 0))).asUInt
```
No need to `.asUInt` here, this actually is a IDEA bug.

sequencer commented on tilelink/src/bundle/TLLink.scala:
```diff
@@ -0,0 +1,93 @@
+
+import scala.collection.immutable.SeqMap
+
+object TLLink {
```
maybe this should be a `utils` object

sequencer commented on tilelink/src/xbar/TLCrossBarParameters.scala:
```diff
@@ -0,0 +1,87 @@
+
+import chisel3.util.log2Ceil
+import upickle.default.{macroRW, readwriter, ReadWriter => RW}
+
```
We do need `ScalaDoc` for this file!

CircuitCoder committed:
Basic crossbar logic

CircuitCoder committed:
Fixed compilation issues

ashen0xa committed:
[misc] add .scalafmt.conf and reformat code

ashen0xa committed:
[misc] automate submodule bump by dependabot

sequencer committed:
[test] add a simple elaborator for testing

sequencer committed:
[test] test will use config.json in os.pwd for now

CircuitCoder committed:
WIP: Master

ashen0xa committed:
[rtl] flip ports correctly

ashen0xa committed:
[misc] add test to generate config json

CircuitCoder committed:
Basic slave / master / simulator

CircuitCoder committed:
Basic direct connection

CircuitCoder committed:
1-x-1 testcase

CircuitCoder committed:
Subtree model external repo (commit 'cc920990bd6f9d85975a413e345b71d529b89f4b')

git-subtree-dir: models
git-subtree-mainline: 6397d79ceb43868e0bbdf2d9cf13381ee402fba6
git-subtree-split: cc920990bd6f9d85975a413e345b71d529b89f4b

CircuitCoder committed:
Request loggers

CircuitCoder committed:
Encapsulate build environment in Nix

CircuitCoder committed:
Add feeding and comparing helper functions
